src/lib/completion.ts

```
@@ -1,131 +1,144 @@
-		return history.map((message: { content: string; type: ChatHistoryType }) => {
-			if (message.type == ChatHistoryType.AI) {
-				return new AIMessage({ content: message.content });
-			} else {
-				return new HumanMessage({ content: message.content });
+	private formatMessages = async (values: InputValues) => {
+		const tools = getTools();
+		const intermediateSteps = values.intermediate_steps
+			? (values.intermediate_steps as Array<AgentStep>)
+			: [];
+		const agentScratchpad = formatLogToString(intermediateSteps);
+		let toolString = '';
+		let toolNames = '';
+		for (let i = 0; i < tools.length; i++) {
+			toolString += '\n' + tools[i].name + ': ' + tools[i].description;
+			toolNames += tools[i].name + ', ';
+		}
+		toolNames = toolNames.slice(0, -1);
+		toolNames = toolNames.slice(0, -1);
+		toolString += '\n\n';
+
+		const formatted = [
+			PREFIX,
+			toolString,
+			TOOL_INSTRUCTIONS_TEMPLATE.replace('{tool_names}', toolNames),
+			SUFFIX.replace('{input}', values.input),
+			agentScratchpad
+		].join('');
+		if (this.verbose) {
+			console.log(colors.fg.green, 'Model prompt: ', colors.style.reset);
+			console.log(colors.fg.blue, formatted, colors.style.reset);
+			console.log('\n');
+		}
+		return [new HumanMessage(formatted)];
+	};
+	private customOutputParser(text: AIMessageChunk): AgentAction | AgentFinish {
+		const content = text.lc_kwargs.content;
+		if (content.includes('Final Answer:')) {
+			if (this.verbose) {
+				console.log(colors.fg.red, 'Model response: ', colors.style.reset);
+				console.log(colors.fg.yellow, content, colors.style.reset);
 			}
-		});
+			const parts = content.split('Final Answer:');
+			const input = parts[parts.length - 1].trim();
+			const finalAnswers = { output: input };
+			return { log: content, returnValues: finalAnswers };
+		}
+		const match = /Action: (.*)\nAction Input: (.*)/s.exec(content);
+		if (match == null) {
+			console.warn('Could not parse output');
+			console.warn(text);
+			process.exit(42);
```

We should't ever exit from here, we need to make sure the backend stays up. Even if we can't continue with the completion I want us to just propagate our errors and give them to the user so they could submit a bug report.

---

You don't need to show errors in any special way to the users, I just want to make sure its returned to the frontend.

---

What should I return then?

---

I would throw an error when this happens and catch it up in the stack.

---

It should be fixed now